### PR TITLE
Crunchbase to elasticsearch

### PR DESCRIPTION
**Pull request by russwinch**
Luigi batch task to load the crunchbase organisation data to elasticsearch
----

Most of the logic is held in the prepare and run steps, but the following components are also included:
- Adjustment to the schema transformer to allow individual dictionaries to be transformed (there is also some tech debt here as I have left the decorator version to get out of sync 👎  as I've not needed to use it)
- Added utilities for batching to misc_utils
- Adjusted existing schema mapping
- Elasticsearch mapping and settings for the new index

I wasted quite a lot of time with the categories but believe this is a fairly efficient way to load them in (compared with the alternatives I tried anyway).

While writing this task another couple of related requirements were discovered and are also part of this PR:

Health labelling
----
This Luigi task collects models from s3 and then applies them to each organisation's categories to determine if it is health related.
An integration test is provided to train and test the model, although the chosen metric *accuracy* is questionable.

Parent ID
----
This should be part of the organisation task but was missed and is therefore added as it's own Luigi task. No new packages were required as it is simple.

Random/additional
----
The requirements file generated by pipreqs is now sorted into alphabetical order, so it's clearer in git diff what has actually been changed. Previously it changed order each time it was generated.

### CODE REVIEW

File: nesta/packages/crunchbase/predict/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+training_data.pickle
```
Comment (jaklinger) [inline nesta/packages/crunchbase/predict/.gitignore]: could you move this to the base .gitignore and use `*pickle`, since now we're working with models we don't want to accidentally upload anything we're going to regret...

File: nesta/packages/crunchbase/predict/src/health_keywords.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+health_keywords = [
```
Comment (jaklinger) [inline nesta/packages/crunchbase/predict/src/health_keywords.py]: store this as a data file, since 

1) It's data 😄 
2) there's a chance that we might want to reproduce this for non-health terms

File: nesta/packages/crunchbase/tests/test_integration_crunchbase_health_predict.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+
+def test_train_and_predict():
+    # collect training data
+    train_data = pd.read_csv(TRAINING_DATA)
+    assert train_data.shape == (5000, 2), 'Training data not shaped as expected'
+
+    # train and check accuracy
+    vec, gs, accuracy = train(train_data)
+    assert accuracy == 0.997
```
Comment (jaklinger) [inline nesta/packages/crunchbase/tests/test_integration_crunchbase_health_predict.py]: Maybe `>= 0.997` would be good, since we're happy for the model to get better

File: nesta/packages/crunchbase/utils.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+def split_str(text):
```
Comment (jaklinger) [inline nesta/packages/crunchbase/utils.py]: I would drop this in favour of `split(',')`

File: nesta/packages/misc_utils/batches.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+"""Utilties for working with batches."""
+import boto3
+import json
+import time
+
+
+def split_batches(data, batch_size):
```
Comment (jaklinger) [inline nesta/packages/misc_utils/batches.py]: Nice!

File: nesta/packages/crunchbase/predict/src/model.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.metrics import confusion_matrix
+from sklearn.model_selection import GridSearchCV, train_test_split
+
+from nesta.packages.crunchbase.utils import split_str
+
+
+def train(data, random_seed=42):
+    """Transform and train.
```
[the remark this answers is not in the document]
Comment (russwinch) [reply nesta/packages/crunchbase/predict/src/model.py]: is this acceptable?
    """Consumes a training dataset, with binary labels and trains a vectoriser and
    random forests classifier."""

I also added a docstring to the top of the file to clarify what it is for, not sure if these are accurate

File: nesta/packages/crunchbase/predict/data.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    Returns:
+        (:obj: `pandas.DataFrame): original data with label added as an additonal column
+    """
+    np.random.seed(random_seed)
+
+    data[label_name] = data[column].apply(lambda x: int(any(k in str(x)
+                                                            for k in keywords)))
+
+    data_idx = np.random.choice(data.index, size, replace=False)
```
Comment (russwinch) [inline nesta/packages/crunchbase/predict/data.py]: I've added `replace=False` here as while adding a test case I discovered the randomiser could use the same data item multiple times. I appreciate that this is the nature of randomness, but are there any negative consequences of this addition? Also, I presume `np.random.choice` is used over `pd.DataFrame.sample` because it is more efficient?

File: nesta/packages/crunchbase/tests/integration_test_dataset_to_label.json [hunk 8]
```diff
@@ -0,0 +1 @@
...we-lot4opqvqc6h", "categories": "building material,construction"}, {"id": "pwkbavmg-37le-nji7-7dy0-1o11eqpg72do", "categories": "banking,finance,financial services"}, {"id": "zc8ql3ox-d616-9gsx-xig8-c7z7wl5fegib", "categories": "consumer goods,digital marketing,e-commerce,fashion,lifestyle,retail,sporting goods,sports"}, {"id": "z1jnmfvg-3uv8-yz1w-h3sp-qkmczn151i9f", "categories": "software,telecommunications"}, {"id": "k764polo-5c72-1qrj-1fi3-pzkk132uw54g", "categories": "information technology"}, {"id": "rwswnin6-5t3m-ml65-dnkc-jg5736a1dosm", "categories": "adventure travel,mobile apps,travel"}, {"id": "n9xac5xz-8x0r-3a1h-1nz6-577t4m6l3ss4", "categories": "manufacturing"}, {"id": "ixo6wp6s-9zqs-87rj-q66b-4j12f7pk9m0w", "categories": ""}, {"id": "jjk4kjyz-unct-5ikv-b9xv-ans9hkkz5n2m", "categories": "internet"}, {"id": "1rrgbg0i-o4wy-tonh-vi4f-o5rp1u2utt8z", "categories": "health care,manufacturing,medical"}, {"id": "m2w8cojd-8ugm-ws0o-u8xn-2w6h6a5a2dd3", "categories": "furniture,home decor,manufacturing"}, {"id": "tl2sza8a-4zzv-wllo-xbhf-c05gjgqjjj0o", "categories": "health care,medical,pharmaceutical"}, {"id": "q0syk9vf-5966-4ckj-gaoj-jlt5tfwvuz9t", "categories": "social media"}]
```
[the remark this answers is not in the document]
Comment (russwinch) [reply nesta/packages/crunchbase/tests/integration_test_dataset_to_label.json]: reduced size of data down to 30 rows

[the remark this answers is not in the document]
Comment (russwinch) [reply nesta/packages/crunchbase/tests/integration_test_dataset_to_label.json]: tar was too much trouble with temp files after the extraction to clean up etc

File: nesta/packages/crunchbase/tests/integration_test_expected_labeled_dataset.json [hunk 9]
```diff
@@ -0,0 +1 @@
...m-l966-onj3-folv6halmhdq", "is_health": false}, {"id": "bov6f5jm-z5we-wfag-ml9b-7msxrf00tmug", "is_health": false}, {"id": "kk2o9x39-otmb-m3vr-fadz-b6evyrl5fomy", "is_health": true}, {"id": "47x1naum-idnz-jzmd-gx9o-d11qxxue1mp6", "is_health": false}, {"id": "2hyb6tq9-mphd-owi1-c7d9-b2kirj89fkwy", "is_health": false}, {"id": "dbx62pft-wywo-763n-51we-lot4opqvqc6h", "is_health": false}, {"id": "pwkbavmg-37le-nji7-7dy0-1o11eqpg72do", "is_health": false}, {"id": "zc8ql3ox-d616-9gsx-xig8-c7z7wl5fegib", "is_health": false}, {"id": "z1jnmfvg-3uv8-yz1w-h3sp-qkmczn151i9f", "is_health": false}, {"id": "k764polo-5c72-1qrj-1fi3-pzkk132uw54g", "is_health": false}, {"id": "rwswnin6-5t3m-ml65-dnkc-jg5736a1dosm", "is_health": false}, {"id": "n9xac5xz-8x0r-3a1h-1nz6-577t4m6l3ss4", "is_health": false}, {"id": "ixo6wp6s-9zqs-87rj-q66b-4j12f7pk9m0w", "is_health": false}, {"id": "jjk4kjyz-unct-5ikv-b9xv-ans9hkkz5n2m", "is_health": false}, {"id": "1rrgbg0i-o4wy-tonh-vi4f-o5rp1u2utt8z", "is_health": true}, {"id": "m2w8cojd-8ugm-ws0o-u8xn-2w6h6a5a2dd3", "is_health": false}, {"id": "tl2sza8a-4zzv-wllo-xbhf-c05gjgqjjj0o", "is_health": true}, {"id": "q0syk9vf-5966-4ckj-gaoj-jlt5tfwvuz9t", "is_health": false}]
```
[the remark this answers is not in the document]
Comment (russwinch) [reply nesta/packages/crunchbase/tests/integration_test_expected_labeled_dataset.json]: reduced to 30 rows

File: nesta/packages/crunchbase/tests/integration_test_training_data.csv [hunk 10]
```diff
@@ -0,0 +1,5001 @@
+category_list,is_Health
```
[the remark this answers is not in the document]
Comment (russwinch) [reply nesta/packages/crunchbase/tests/integration_test_training_data.csv]: reduced from 5000 to 100 rows of data to train on. accuracy around 90% with such little data, but it it reproducible, which is the intention of the integration test...and runs really fast
